File: CTCI-2026/Ch02_Linked_Lists/q2_2_kth_to_last.py
# ruff & black
import unittest


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def create_linked_list(values):
    head = None
    tail = None
    for value in values:
        node = Node(value)
        if head is None:
            head = node
            tail = node
        else:
            tail.next = node
            tail = node
    return head


def kth_to_last(head, k):
    if k <= 0:
        return None

    first = head
    second = head

    for _ in range(k):
        if second is None:
            return None
        second = second.next

    while second is not None:
        first = first.next
        second = second.next

    return first


class TestKthToLast(unittest.TestCase):
    def test_k_greater_than_length(self):
        head = create_linked_list([1, 2, 3])
        node = kth_to_last(head, 4)
        self.assertIsNone(node)

    def test_empty_list(self):
        head = None
        node = kth_to_last(head, 1)
        self.assertIsNone(node)

    def test_single_node(self):
        head = create_linked_list([55])
        node = kth_to_last(head, 1)
        self.assertIsNotNone(node)
        self.assertEqual(node.data, 55)

    def test_overloads_const_and_non_const(self):
        head = create_linked_list([1, 2, 3, 4, 5])

        mutable_result = kth_to_last(head, 2)
        self.assertIsNotNone(mutable_result)
        self.assertEqual(mutable_result.data, 4)

        const_head = head
        const_result = kth_to_last(const_head, 2)
        self.assertIsNotNone(const_result)
        self.assertEqual(const_result.data, 4)

    def test_boundaries(self):
        head = create_linked_list([10, 20, 30, 40])

        tail = kth_to_last(head, 1)
        self.assertIsNotNone(tail)
        self.assertEqual(tail.data, 40)

        first = kth_to_last(head, 4)
        self.assertIsNotNone(first)
        self.assertEqual(first.data, 10)

    def test_sample(self):
        head = create_linked_list([1, 2, 3, 4, 5])
        node = kth_to_last(head, 2)
        self.assertIsNotNone(node)
        self.assertEqual(node.data, 4)


if __name__ == "__main__":
    unittest.main()

File: CTCI-2026/Ch02_Linked_Lists/q2_2_kth_to_last.cpp
#include <gtest/gtest.h>

#include "../common/linked_list_utils.hpp"

/** Returns the kth to last node in a singly linked list.
 *
 * Single-pass two-pointer approach is optimal: O(n) time, O(1) extra space.
 */
template <typename NodePtr>
[[nodiscard]] NodePtr kth_to_last_impl(NodePtr head, size_t k) noexcept {
    if (!head) return nullptr;

    auto distance = k;
    auto second = head;
    for (; second && distance; second = second->next.get(), --distance) {
        // Empty body is clearer than a trailing semicolon
    }
    if (!second) {
        if (distance == 0) {
            // k is exactly the length of the list, so we return the head
            return head;
        }
        return nullptr;
    }
    auto first = head;
    for (; second; first = first->next.get(), second = second->next.get()) {
        // Empty body is clearer than a trailing semicolon
    }
    return first;
}

Node* kth_to_last(Node* head, size_t k) noexcept {
    return kth_to_last_impl(head, k);
}

const Node* kth_to_last(const Node* head, size_t k) noexcept {
    return kth_to_last_impl(head, k);
}

TEST(test_limits, k_greater_than_length) {
    auto list = create_linked_list({1, 2, 3});
    auto node = kth_to_last(list.get(), 4); // k is larger than the list
    EXPECT_EQ(node, nullptr); // Or however you chose to handle errors
}

TEST(test_limits, empty_list) {
    std::unique_ptr<Node> list;
    auto node = kth_to_last(list.get(), 1);
    EXPECT_EQ(node, nullptr);
}

TEST(test_limits, single_node) {
    auto list = create_linked_list({55});
    auto node = kth_to_last(list.get(), 1);
    ASSERT_NE(node, nullptr);
    EXPECT_EQ(node->data, 55);
}

TEST(test_overloads, const_and_non_const) {
    auto list = create_linked_list({1, 2, 3, 4, 5});
    auto* mutable_result = kth_to_last(list.get(), 2);
    ASSERT_NE(mutable_result, nullptr);
    EXPECT_EQ(mutable_result->data, 4);

    const Node* const_head = list.get();
    auto const_result = kth_to_last(const_head, 2);
    ASSERT_NE(const_result, nullptr);
    EXPECT_EQ(const_result->data, 4);
}

TEST(test_limits, boundaries) {
    auto list = create_linked_list({10, 20, 30, 40});

    // k = 1: The tail
    auto tail = kth_to_last(list.get(), 1);
    EXPECT_EQ(tail->data, 40);

    // k = length: The head
    auto head = kth_to_last(list.get(), 4);
    EXPECT_EQ(head->data, 10);
}

TEST(test_sample, test1) {
    auto list = create_linked_list({1, 2, 3, 4, 5});
    auto node = kth_to_last(list.get(), 2);
    ASSERT_TRUE(node);
    EXPECT_EQ(node->data, 4);
}
